Move import queue out of network

gterzian opened the pull request:
FIX #1168

gterzian commented:
@rphmeier Ok so at this point, it's basically a huge mess, but it compiled locally...

One problem I ran into was the usage of `client::Error` inside `BlockImport`. It's mainly a problem because at this point, using `client` from `consensus-common` creates a circular dependency. Also, this `client::Error` is basically that which is handled [here](https://github.com/paritytech/substrate/blob/638125e79dfe1aeefcdb69f0a55abf6854925be9/core/consensus/common/src/import_queue.rs#L402), so it might actually make sense to switch it to a "consensus error", since it's really about the import queue asking the client to import the block. Or perhaps there is something that could be done using the "chain link"?

So at this point, I move over `ImportQueue` and `BasicQueue`, as well as `Link`, and left `NetworkLink` inside "network". 

Next step, clean it up(this was basically a race to get the changes to compile), and after that I'd like to see if the network -> consensus comm can be done using a channel(although it might actually better to do it in a separate PR). 

Preliminary comments welcome, I need to take a break...

gterzian commented on core/consensus/common/src/import_queue.rs:
```diff
@@ -69,7 +79,7 @@ pub trait ImportQueue<B: BlockT>: Send + Sync {
 	/// This is called automatically by the network service when synchronization
 	/// begins.
-	fn start<L>(&self, _link: L) -> Result<(), Error> where
+	fn start<L>(&self, _link: L) -> Result<(), std::io::Error> where
```
Here I switched from `client::Error` to a plain `io::Error`, since this is really about starting the import thread successfully...

gterzian commented on core/consensus/common/src/import_queue.rs:
```diff
@@ -34,17 +34,27 @@ use primitives::AuthorityId;
+
+/// Block data used by the queue.
+#[derive(Debug, PartialEq, Eq, Clone)]
+pub struct IncomingBlock<B: BlockT> {
```
This is essentially the consensus's view into `BlockData`

gterzian commented on core/consensus/common/src/import_queue.rs:
```diff
@@ -354,10 +317,10 @@ enum BlockImportError {
 
 /// Import a bunch of blocks.
 fn import_many_blocks<'a, B: BlockT, V: Verifier<B>>(
-	import_handle: &BlockImport<B, Error=ClientError>,
```
The main point of this was to avoid the circular deps that arises if you use "Client" in here...

gterzian commented on core/finality-grandpa/src/lib.rs:
```diff
@@ -847,22 +847,24 @@ impl<B, E, Block: BlockT<Hash=H256>, RA, PRA> BlockImport<Block>
+		);
+
+        if maybe_change.is_err() {
+            return Err(ConsensusErrorKind::ClientImport.into())
```
These `is_err` are a temporary hack to see if I could get the whole thing to compile, I guess some `Into` will need to be implemented instead...

gterzian commented:
What is your convention with line spacing? I see a few indents came off, and I can't see it immediately on my editor...

gterzian commented:
Looks like I also need to update the tests...

bkchr commented:
@gterzian for indenting we use tabs.

bkchr commented on core/finality-grandpa/src/lib.rs:
```diff
@@ -847,22 +847,24 @@ impl<B, E, Block: BlockT<Hash=H256>, RA, PRA> BlockImport<Block>
-		)?;
+		);
+
+        if maybe_change.is_err() {
```
Spaces

bkchr commented on core/finality-grandpa/src/lib.rs:
```diff
@@ -905,17 +911,27 @@ impl<B, E, Block: BlockT<Hash=H256>, RA, PRA> BlockImport<Block>
-		})?;
+		});
+
+        if import_result.is_err() {
```
Spaces.

gterzian commented on core/network/src/protocol.rs:
```diff
@@ -393,19 +393,8 @@ impl<B: BlockT, S: NetworkSpecialization<B>, H: ExHashT> Protocol<B, S, H> {
-
-
+		let mut sync = self.sync.write();
+		sync.on_block_data(&mut ProtocolContext::new(&self.context_data, io), peer, request, response)
```
Ok, so, like the comment above read, this is indeed deadlocking :)

Could be something that goes away using channels, which I'll try later, so leaving this for now...

bkchr commented on core/finality-grandpa/src/lib.rs:
```diff
@@ -872,9 +874,13 @@ impl<B, E, Block: BlockT<Hash=H256>, RA, PRA> BlockImport<Block>
-				)?;
+				);
+
+                if tree_route.is_err() {
```
I did not mark all space usages. Here are more. Could you please check all of your changes :)

gterzian replied:
Will do!

bkchr commented on core/finality-grandpa/src/lib.rs:
```diff
@@ -905,17 +911,27 @@ impl<B, E, Block: BlockT<Hash=H256>, RA, PRA> BlockImport<Block>
-		})?;
+		});
+
+		if import_result.is_err() {
```
I would propose to change this if and down to line 925 to:
```
match import_result {
    Ok(ImportResult::Queued) => {},
    Ok(r) => return Ok(r),
    Err(_) => return Err(ConsensusErrorKind::ClientImport.into()),
};
```

bkchr commented on core/finality-grandpa/src/lib.rs:
```diff
@@ -905,17 +911,27 @@ impl<B, E, Block: BlockT<Hash=H256>, RA, PRA> BlockImport<Block>
-		})?;
+		});
+
+        if enacts_change.is_err() {
```
Here You can do the same:
```
match enacts_change {
   Ok(false) => return Ok(import_result),
   ....
```

gterzian replied:
Thanks, I will revisit this code...

gterzian commented:
`test tests::authoring_blocks ... test tests::authoring_blocks has been running for over 60 seconds`

Ok, looks like it's building, and deadlocking because of the first commit.

TODO:

1. Major clean-up, spaces to  tabs, unused imports, and so on...
2. Try to change the import thread to using channels(which should fix the deadlock as well).

gterzian commented:
This one is ready for review. Couple of recaps:

1. In order to prevent a circular dependency when using "client" inside "consensus-common", I switched `BlockImport` to using an error type from consensus.
2. Because the tests are using "client" as well, I left those in network, and moved them to a separate file. I guess this makes sense since these are in some way "integration" tests involving network, client, and consensus...

gterzian commented:
@rphmeier Ok, comments have been addressed(in various commits to be squashed).

On the error handling, since the import queue doesn't actually do anything but trace the errors [here](https://github.com/paritytech/substrate/blob/044b054023f22867b0e4f24cccfb4c5df9154311/core/consensus/common/src/import_queue.rs#L422), I though it might be enough for now to add the "other error" as a string inside the `ConsensusError::ClientImport` one. What do you think?

rphmeier commented:
Sounds fine for now

gavofyork commented:
@rphmeier time to update your review?

rphmeier commented:
Thanks!
